### PR TITLE
🐛 bug(Airbyte CDK): fix case where stream wont have a state attribute and needs to resolve get_updated_state

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What
Several customers have reported an increase in billing. After researching, it appears that each connection has sharply increased data transfer. Further research shows that the syncs have been reloading data since the 9th and appending it without updating states. This is related to the [CDK upgrade](https://github.com/airbytehq/airbyte/pull/41044/files) in stripe and [this change in CDK](https://github.com/airbytehq/airbyte/pull/37429/files#diff-2a7a4277f09aa3acd3f25ae628f9dfd876505921f9c789e96c8666a96c1bce79).

Basically:
- Context: source-stripe gets records in descending order (recent to oldest).
- Before 0.87.0, the stream state was updated from the first record and not again, as the next records had min values.
- After 0.87.0, the original state is never updated (at least the variable that holds the value, I mean), and the next record cursors always update the checkpoint-reader internal state as they are always greater than the initial state, even if the cursor is minor compared to the previous record. So, in the end, the checkpoint is really close to the initial state.

So we need to fix this for the state to get updated correctly from the most recent record.

Resolves  [/issues/5865](https://github.com/airbytehq/oncall/issues/5865)

## How
Assign the `get_updated_data` result to state in every iteration of the records read in the base stream.

## Review guide

1. `airbyte-cdk/python/airbyte_cdk/sources/streams/core.py`: Assign the `get_updated_data` result to state in every iteration of the records read in the base stream.
2. `airbyte-cdk/python/unit_tests/sources/test_abstract_source.py`: new unit test to cover this functionality

## User Impact
Once we prepare a new PR to upgrade the CDK in Stripe, we can unpin the old version of the connector to use this solution, and no increasing syncs should be observed.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/python/unit_tests/sources/test_abstract_source.py [hunk 1]
```diff
@@ -1297,6 +1303,71 @@ def test_emit_non_records(self, mocker):
 
         assert messages == expected
 
+    @pytest.mark.parametrize(
+        "use_legacy",
```
Comment (brianjlai) [inline airbyte-cdk/python/unit_tests/sources/test_abstract_source.py:1307]: I would actually say we can drop this `use_legacy` scenario in this test. Legacy state is rarely used anymore and it doesn't impact the root scenario of what the new test is trying to validate anyway

**Review by brianjlai — APPROVED** [review]
this is great aldo!

File: airbyte-cdk/python/unit_tests/sources/test_abstract_source.py (hunk 1 above)
Comment (aldogonzalez8) [reply airbyte-cdk/python/unit_tests/sources/test_abstract_source.py]: Aaaaaand is gone.

**Review by bazarnov — APPROVED** [review]

**Comment by aldogonzalez8** [comment]
/approve-regression-tests

[Check job output.](https://github.com/airbytehq/airbyte/actions/runs/9991656454)

**Comment by aldogonzalez8** [comment]
[Here](https://ff1dc771106eeb2a40756cb884e0f4d84cd0cfe8568d6bbb9c660e1-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Ftest%2Fmanual%2Fmaster%2F1721306391%2F3cbd292154a8b8e942348e079404ae3a5146181c%2Fsource-shopify%2F2.4.13%2Foutput.html?jk=AYBlUPDsOq4V58d0Dia50zxZLlMD6AQYXq3utgsRbU-vaRFu8FLwx3rrmDnE7QlQv8dHzpk8An2GfN0CnkVy7fR07uvOfymPD-3pfsZLtP2MYdWLDYtAEoMl94QxBpUPY1H1Cx6pnyCL3TS5k7PDU7u6rpTuTa54Wqog7Y7zpz2iaDE4NOpQO44FScp3gtlP3KwAk9R0uVzDoxBiWOMxkt3WeoIDdMtmqFQjKmcpMPINVHC6rrNGMKWCGbDy1hKCfEUFA7T9freucehi2PMTwZHA8R6CpJ6WvCwYVGk3e96Pq4CXWZ8-PSyrBuMmn_CgvW6ESAlnQH39BMCnIH0ervb0XQ68W8PMczXZcesCgkouHhI27Cw3ZbWaFhoGuEVQfxRcboDRavN8q35InqBRprVanNk5M8TP7aMdh5dfESDja64z0uJTvlXFRkRHUqbP7OAfVK1pqabph3lFfUwq69W5ex9plAMo2KNkmcuas6AbSQby-Fcj9YzYJvLTIstaHiX2YQkqFw0QFHXFmwMZUvwpx4GxVAXEioA4weyBTGtLC4MU7pkzBZtfSMmvnkwUoiAXZsN81LiifaSPL5aF52IEzC5dMe7SlirfRvn83KSw2ie9bpGY5qN51LyquFjk-ktu515WYvQYmCCCtwjaDXcDA4ZYwV8nUuSTU_-IdnhSK3mwwN8D3z3vAsZPitYz4t794hntHNCLcKkwvdw6VeDghoanjKwPmM4-rFyZxK34o5_ryqS-Syv5V5I3eR3ccwwbMJxxUYhsrNez_PDFgn1WPGortWyXTzgfev23i3cHO0RmMm-dDmBqWUBe2-CgNJY2GUH7LS7kRgxDCcaLBtyn8VLisdk8bOMo2pTt8M6VWu3c43y85eB-F0sa1C3s5b_VGSvGtwepRyL0c2s76daZ9SsMdDCk_UiH7SFn-nCKVKNhCvrXIOnj9Xd-BF6kubUN_zWup0Hf_j3x8OhHVGWQT6Th9bm7pE7wb5gJUHjj7AY469VY_PxYlwe7Emj5ssiMvrrWxvaUBRjFdFctWX0tjZJhXN3yxTQ6Iwun83GJm3Nj7JWLn6B9OLyclXD7TBOxPygR1ltK06QwV8s1441ilh7ZwR0LMgNqp5zNkpoo5O0W0DBafUJ005FOyP-PsGATyI3z-sLmHFUaKxAl79r-pXNxYA2mIV5c67T8bVsd05T0IALkM2SgMbDe4rosB7urw7Wm3_ZYCMlOsjmV0f7KKtU0FdCe0Cvi10MAF2Y-O7kJ0kLzI0fxCOOeea6ZeXFL5YruuyrXT9pkVfAIj3sBDTUNNTtVuQ8MG-UT5raa8ZUlIjmQ42Pi9yP6NCk6tXXgDzBThKH5N2pA&isca=1) you can find the results of the regression tests

**Comment by aldogonzalez8** [comment]
Current concern: making this change will make source-shopify change how it is checkpoint right now.

For example, from test_read_sequential_slices CAT when the initial state is {} will change state messages from this:
```
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2021-06-23T01:09:58+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 250.0}, 'destinationStats': None}, 'trace': None, 'control': None}
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2021-06-23T01:10:30+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 250.0}, 'destinationStats': None}, 'trace': None, 'control': None}
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2021-06-23T01:11:03+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 250.0}, 'destinationStats': None}, 'trace': None, 'control': None}
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2023-12-11T10:37:41+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 172.0}, 'destinationStats': None}, 'trace': None, 'control': None}
```

To this:
```
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2023-04-13T12:00:55+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 250.0}, 'destinationStats': None}, 'trace': None, 'control': None}
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2023-04-13T12:00:55+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 250.0}, 'destinationStats': None}, 'trace': None, 'control': None}
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2023-04-13T12:09:45+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 250.0}, 'destinationStats': None}, 'trace': None, 'control': None}
{'type': 'STATE', 'log': None, 'spec': None, 'connectionStatus': None, 'catalog': None, 'record': None, 'state': {'type': 'STREAM', 'stream': {'stream_descriptor': {'name': 'inventory_levels', 'namespace': None}, 'stream_state': {'updated_at': '2023-12-11T10:37:41+00:00'}}, 'global_': None, 'data': None, 'sourceStats': {'recordCount': 172.0}, 'destinationStats': None}, 'trace': None, 'control': None}

```

And then the job fails as running read with state messages affect expected count:
len records_N
4

expected_records_count >=
172.0

**Comment by bazarnov** [comment]
Aldo, could you please test your changes on this Shopify PR?

https://github.com/airbytehq/airbyte/pull/42095

I've added the `default ASC sorting` to all Bulk Streams, it's expected we have the checkpointing work correctly with this `Shopify change` and this should guarantee the order of records for the future because we don't care now about how the records were ordered from the part of the API (BULK streams), we make the nested sorting for collected records now, based on the `checkpoint_interval`. It means that once we hit the `checkpoint_interval`, let's say 250 records collected, we `sort` them by the `stream cursor field` and then `emit` in `ASC` order.

**Comment by bazarnov** [comment]
The `shopify` CAT fails due to the non-ordered records expected to be ordered. This is fixed here: https://github.com/airbytehq/airbyte/pull/42095

**Comment by aldogonzalez8** [comment]
> The `shopify` CAT fails due to the non-ordered records expected to be ordered. This is fixed here: #42095

```
Testing Scenario steps description:
git checkout baz/source/shopify/fix-base-url-mismatch
git merge aldogonzalez8/fix_state_not_updating_for_stream_with_records_desc_order
airbyte-ci connectors --name=source-shopify --use-local-cdk test --only-step="acceptance"
```

image

**Comment by aldogonzalez8** [comment]
The Zendesk cat is failing because of an invalid record. This is fixed here: https://github.com/airbytehq/airbyte/pull/42455

**Pull request merged**